### PR TITLE
chore(deps): Upgrade @typescript-eslint/*

### PR DESCRIPTION
**Pull request by sirreal**
#### Changes proposed in this Pull Request

* Upgrade @typescript-eslint dependencies.
* Upgrade `eslint-plugin-jsdoc`

After #46198, TypeScript files started showing eslint apparently due to incompatibilities with TypeScript v4. Eslint would error with `Parsing error: Cannot read property 'map' of undefined`.

After upgrading `@typescript-eslint` scope, the jsdoc plugin started erroring, so I upgraded that too.

`@typescrit-eslint` 4.0.0 includes many [breaking changes](https://github.com/typescript-eslint/typescript-eslint/releases/tag/v4.0.0), one of which affect us: they forked `eslint-scope` to support TypeScript and provide better info for rules that depend on the scope, like `no-shadow`. They then [re-implemented those rules](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/README.md#extension-rules) so they understand TypeScript. Long story short, if we run eslint rules like `no-shadow` in TypeScript files they will fail. Based on the list in [their doc](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/README.md#extension-rules) I have disabled all eslint rules affected by this change.

#### Concerns

I upgraded the jsdoc plugin as the simplest fix for errors. I'm not sure [what may have changed](https://github.com/gajus/eslint-plugin-jsdoc/compare/v18.11.0...v30.6.3) in its defaults. It doesn't seem to include a changelog.

#### Testing instructions

* Linting TypeScript files on the main branch will error (the linter errors, not produced lint errors).
* Linting is fixed on this branch.

### CODE REVIEW

**Comment by sirreal** [comment]
Related: https://github.com/typescript-eslint/typescript-eslint/issues/2098

**Comment by sirreal** [comment]
> I upgraded the jsdoc plugin as the simplest fix for errors. I'm not sure [what may have changed](https://github.com/gajus/eslint-plugin-jsdoc/compare/v18.11.0...v30.6.3) in its defaults. It doesn't seem to include a changelog.

The releases include a changelog and can be read starting [here](https://github.com/gajus/eslint-plugin-jsdoc/releases?after=v18.11.0).

**Comment by scinos** [comment]
*Updated*: @typescrit-eslint 4.0.0 includes many breaking changes, one of which affect us: they forked eslint-scope to support TypeScript and provide better info for rules that depend on the scope, like no-shadow. They then re-implemented those rules so they understand TypeScript. Long story short, if we run eslint rules like no-shadow in TypeScript files they will fail. Based on the list in their doc I have disabled all eslint rules affected by this change.

Before disabling these rules, the set of errors was:

```
721 files checked.  326 passed.  395 failed.  582 warnings.  480 errors.
? Which rule would you like to fix? (Use arrow keys)
❯ @typescript-eslint/ban-ts-comment:                   3|
  @typescript-eslint/ban-ts-ignore:                    3|
  @typescript-eslint/ban-types:                      102|
  @typescript-eslint/explicit-module-boundary-types: 512|
  @typescript-eslint/no-empty-function:                2|
  @typescript-eslint/no-explicit-any:                 55|
  @typescript-eslint/no-inferrable-types:              2|
  @typescript-eslint/no-unused-vars:                   1|
  @typescript-eslint/no-use-before-define:            81|
  import/no-extraneous-dependencies:                   3|
  jsdoc/check-param-names:                             2|
  jsdoc/newline-after-description:                     3|
  jsdoc/require-param:                                10|
  jsdoc/require-property-description:                  2|
  no-duplicate-imports:                                1|
  no-redeclare:                                       14|
  no-restricted-imports:                               1|
  no-shadow:                                          23|
  no-undef:                                          229|
  react-hooks/exhaustive-deps:                        13|
```

After disabling the extension rules counterpart, it is:

```
721 files checked.  344 passed.  377 failed.  582 warnings.  442 errors.
? Which rule would you like to fix? (Use arrow keys)
❯ @typescript-eslint/ban-ts-comment:                   3|
  @typescript-eslint/ban-ts-ignore:                    3|
  @typescript-eslint/ban-types:                      102|
  @typescript-eslint/explicit-module-boundary-types: 512|
  @typescript-eslint/no-empty-function:                2|
  @typescript-eslint/no-explicit-any:                 55|
  @typescript-eslint/no-inferrable-types:              2|
  @typescript-eslint/no-unused-vars:                   1|
  @typescript-eslint/no-use-before-define:            81|
  import/no-extraneous-dependencies:                   3|
  jsdoc/check-param-names:                             2|
  jsdoc/newline-after-description:                     3|
  jsdoc/require-param:                                10|
  jsdoc/require-property-description:                  2|
  no-restricted-imports:                               1|
  no-undef:                                          229|
  react-hooks/exhaustive-deps:                        13|
```

File: .eslintrc.js [hunk 1]
```diff
@@ -114,14 +114,47 @@ module.exports = {
...
+					'no-shadow': 'off',
+					'no-unused-expressions': 'off',
+					'no-unused-vars': 'off',
+					'no-use-before-define': 'off',
+					'no-useless-constructor': 'off',
+					quotes: 'off',
+					'require-await': 'off',
+					'return-await': 'off',
+					semi: 'off',
+					'space-before-function-paren': 'off',
```
Comment (jsnajdr) [inline .eslintrc.js:149]: Wow, seems that the default `eslint-recommended` is seriously incomplete 🙁

By the way, I don't think we use the `@typescript-eslint/eslint-plugin`s `eslint-recommended` config at this moment, so for 100% correctness, this list should include all rules from there, too. E.g., `no-dupe-args`.

**Review by jsnajdr — APPROVED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (scinos) [reply .eslintrc.js:155]: I added the rules from `@typescript-eslint/eslint-plugin`. The final set of errors is:

```
❯ @typescript-eslint/ban-ts-comment:                   3|
  @typescript-eslint/ban-ts-ignore:                    3|
  @typescript-eslint/ban-types:                      102|
  @typescript-eslint/explicit-module-boundary-types: 516|
  @typescript-eslint/no-empty-function:                2|
  @typescript-eslint/no-explicit-any:                 56|
  @typescript-eslint/no-inferrable-types:              2|
  @typescript-eslint/no-unused-vars:                   1|
  @typescript-eslint/no-use-before-define:            81|
  import/no-extraneous-dependencies:                   3|
  jsdoc/check-param-names:                             5|
  jsdoc/newline-after-description:                     3|
  jsdoc/require-param:                                13|
  jsdoc/require-property-description:                  2|
  no-restricted-imports:                               1|
  react-hooks/exhaustive-deps:                        13|
```

**Review by wp-desktop — CHANGES_REQUESTED** [review]
WordPress Desktop CI Failure for job "wp-desktop-mac".

@scinos please inspect this job's build steps for breaking changes at [this link](https://circleci.com/gh/Automattic/wp-calypso/918570). For temporal failures, you may try to "Rerun Workflow from Failed".

Please also ensure this branch is rebased off latest Calypso.

**Pull request merged**